commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/AbstractChunkIndex.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class AbstractChunkIndex implements ChunkIndex {
+    protected final int originalChunkSize;
+    protected final int originalFileSize;
+    protected final int finalTransformedChunkSize;
+
+    protected AbstractChunkIndex(final int originalChunkSize,
+                                 final int originalFileSize,
+                                 final int finalTransformedChunkSize) {
+        checkSize(originalChunkSize, "Original chunk size");
+        this.originalChunkSize = originalChunkSize;
+
+        checkSize(originalFileSize, "Original file size");
+        this.originalFileSize = originalFileSize;
+
+        checkSize(finalTransformedChunkSize, "Final transformed chunk size");
+        this.finalTransformedChunkSize = finalTransformedChunkSize;
+    }
+
+    // Override with care, this method is indirectly called from a constructor.
+    protected final List<Chunk> materializeChunks() {
+        final List<Chunk> chunks = new ArrayList<>(chunkCount());
```

Comment:
`materializeChunks` is now final but `chunkCount` is not and I guess there is exactly the same issue with a call from constructor here.

Reply:
And AFAIU an option probably could be to make `chunkCount` a final field initialized in constructor instead since all the info is available in constructor to calculate it.

Reply:
That's a good point, I made `chunkCount` a final field